src/libs/ActiveClientManager/index.js

```diff
@@ -6,38 +12,61 @@ import * as ActiveClients from '../actions/ActiveClients';
 
 const clientID = Str.guid();
 const maxClients = 20;
-
-let activeClients;
-
-let resolveIsReadyPromise;
-const isReadyPromise = new Promise((resolve) => {
-    resolveIsReadyPromise = resolve;
+let addedSelf = false;
+let activeClients = [];
+let resolveInitedPromise;
+const initedPromise = new Promise((resolve) => {
+    resolveInitedPromise = resolve;
+});
+let resolveSavedSelfPromise;
+const savedSelfPromise = new Promise((resolve) => {
+    resolveSavedSelfPromise = resolve;
 });
 
 /**
+ * Determines when the client is ready. We need to wait both till we saved our ID in onyx AND the init method was called
  * @returns {Promise}
  */
 function isReady() {
-    return isReadyPromise;
+    return Promise.all([initedPromise, savedSelfPromise]);
 }
 
 Onyx.connect({
     key: ONYXKEYS.ACTIVE_CLIENTS,
     callback: (val) => {
-        activeClients = !val ? [] : val;
-        if (activeClients.length >= maxClients) {
-            activeClients.shift();
+        // We only add this client once, ensuring it is the last in the array (that determines who is leader)
+        if (!addedSelf) {
+            activeClients = _.without(val, clientID);
+            activeClients.push(clientID);
             ActiveClients.setActiveClients(activeClients);
         }
+
+        // Remove from the beginning of the list any clients that are past the limit, to avoid having thousands of them
+        let removed = false;
+        while (activeClients.length >= maxClients) {
+            activeClients.shift();
+            removed = true;
+        }
+
+        // Save the clients back to onyx, if they changed
+        if (removed || !addedSelf) {
+            ActiveClients.setActiveClients(activeClients).then(() => {
+                // If we just added ourselves, we need to resolve the promise so that isReady fires
```

Comment:
This comment doesn't really seem consistent with the logic below. It says that the promise needs resolved if we just added ourselves, but shouldn't that mean that the early return should be `if (!addedSelf)`?

Reply:
Stupid linter made me change this to early return and now is a bit more confusing, but no, both are correct. We only resolve the promise if we have not addedSelf yet, which is what the code is doing

Reply:
OK, that still doesn't make a whole lot of sense to me. That would imply that this callback is called multiple times. Once with `addedSelf === true` and once with `addedSelf === false`, right? But since `ActiveClients.setActiveClients()` is wrapped in an `if (!addedSelf)`, how is it ever called with `addedSelf === true`?

Is the early return even necessary?

Reply:
The callback is only called once, but `isReady` is called many times (each time the sequential queue runs).

> Is the early return even necessary?

hmmm probably not, because resolving a promise multiple times does nothing really... right?

Reply:
Yeah, resolving a promise twice does not trigger anything again (ie. any `.then()`s will not be called)

Reply:
I think I could remove the early return then